### ISSUE TITLE
ClusterSecretStore serviceAccountRef defaults to the namespace of the referent

### ISSUE THREAD

**Issue by moolen** [external-secrets/external-secrets#961]
@alfredkrohmer i just created this issue to track your problem.

> Just discovered this while testing the 0.5.0 release, this change breaks a use case that we have: **not** specifying the namespace in the `secretRef` or `kubernetesServiceAccountToken` sections in the `auth` spec for certain providers (Vault in our case) in `ClusterSecretStore`s. The namespace of the `ExternalSecret` referring to the `ClusterSecretStore` would then be used when authenticating against the provider.
>
> Implementation for `secretRef` in Vault provider:
> https://github.com/external-secrets/external-secrets/blob/787129e3250d49733d1ae8dcd7dc3acd4a557182/pkg/provider/vault/vault.go#L829-L832
> 
> Implementation for `kubernetesServiceAccountToken` in Vault provider:
> https://github.com/external-secrets/external-secrets/blob/787129e3250d49733d1ae8dcd7dc3acd4a557182/pkg/provider/vault/vault.go#L853-L861

_Originally posted by @alfredkrohmer in https://github.com/external-secrets/external-secrets/issues/750#issuecomment-1090150833_

So, the behavior you describe is not intended. The `ClusterSecretStore` should point to a single service account in a particular namespace and not use a SA from the namespace the ExternalSecret is in.

I guess it is this line specifically that shouldn't be there.
https://github.com/external-secrets/external-secrets/blob/787129e3250d49733d1ae8dcd7dc3acd4a557182/pkg/provider/vault/vault.go#L834

I'm interested in how your use-case looks in detail, can you provide more information? Do you basically share one `ClusterSecretStore` that points to a vault and configures things "globally" but the authentication bits lie in each and every namespace?

**Comment by alfredkrohmer** [comment]
> I'm interested in how your use-case looks in detail, can you provide more information? Do you basically share one `ClusterSecretStore` that points to a vault and configures things "globally" but the authentication bits lie in each and every namespace?

Exactly. That way we can allow access in Vault via ACL path templating that refers to the service account namespace.

**Comment by gusfcarvalho** [comment]
Should we standardize this behavior across providers?

**Comment by moolen** [comment]
It's a little to magical/implicit. I'd favor a separate field in the provider crd to make that explicit.

**Comment by alfredkrohmer** [comment]
Actually the documentation of the `Namespace` field in these two types already says that it defaults to the namespace of the referent (which I guess is the `ExternalSecret`?) for cluster-scoped resources:
https://github.com/external-secrets/external-secrets/blob/4cbf1b8742fb1f4279caf513f9c4b2bb90ab5c29/apis/meta/v1/types.go#L19-L40

**Comment by gusfcarvalho** [comment]
I can tell for sure that for the majority of the providers, an error is given if `ClusterSecretStore` is used with a `SecretKeySelector.Namespace == nil`, so this comment is not really reflecting the reality here.. I agree with @moolen in the sense that an user might not just understand what is fully going on, leading to some permissions not being handled properly.

**Comment by alfredkrohmer** [comment]
> One question that I have, though: How does the `ClusterSecretStore` gets `Validate()'d` in this case?
> 
> (If we think about implementing it across all providers, #948 can be a footgun if we do not think of a clear way to Validate the ClusterSecretStore).

It doesn't, it constantly fails validation but the `ExternalSecret` still syncs.

**Comment by moolen** [comment]
:thinking: I can't come up with a solution i like by changing the CRD (e.g. adding a `namespacedServiceAccountRef` that does what is described above). 

Although `defaults to the namespace of the referent` is a little magical i think it's the best solution. This shouldn't be a breaking change. Previously it was ignored by most providers or a error was returned. The only exception seems to be `vault`.

I'd suggest that we actually implement what is documented :joy: :trollface:

WDYT?

**Comment by alfredkrohmer** [comment]
I guess adding a `namespacedServiceAccountRef` might get difficult since the `ClusterSecretStore` uses the same structs as the `SecretStore` for the provider configuration (or it could just be ignored if configured on a `SecretStore`, but that also would look kinda weird).

How about a CLI toggle for the behavior?

**Comment by gusfcarvalho** [comment]
I agree to actually implement it across all the other providers. However, regarding the `floodgate` that we want to put. It would effectively break this approach completely, as the ClusterSecretStore would never be validated. Besides implementing the feature for the other providers, we could probably make vault's `Validate()` method use `utils.ValidateNetwork` instead of checking if the token is valid.

**Comment by gusfcarvalho** [comment]
> How about a CLI toggle for the behavior?
This would be very specific, and a little bit confusing to other users to figure it out.

Since we are going to the direction of actually making vault the standard here, I think we should not need to specify a new CLI option for it.

**Comment by moolen** [comment]
> regarding the floodgate that we want to put

True, it shouldn't actually validate the credentials but use `utils.ValidateNetwork` instead. We must not forget that when implementing it across all providers :exclamation:

**Comment by agaudreault-jive** [comment]
In the meantime, can we get a fix in `0.5.X` to do what the documentation is saying and remediate to the issue until something more bulletproof is implemented in 0.6.0 ?

```
admission webhook "validate.clustersecretstore.external-secrets.io" denied the request: invalid Auth.Kubernetes.ServiceAccountRef: cluster scope requires namespace
```

**Comment by moolen** [comment]
> In the meantime, can we get a fix in 0.5.X to do what the documentation is saying 

yeah, i agree. We can get this into the next patch release i think:
1. disable validation of `secretRef.namespace`
2. add tests specifically for vault to support this use-case

That enables it at least for the vault provider. other providers haven't implemented that AFAIK.

The actual "fix" will be more effort as we need to touch all providers for that.

**Comment by moolen** [comment]
TODOs to implement referent auth:
- [x] HashiCorp Vault #1176 
- [ ] Kubernetes #1201 
- [ ] AWS Secrets Manager
- [ ] AWS Parameter Store
- [ ] Azure KeyVault
- [ ] GCP Secret Manager
